SteamKit2/SteamKit2/Util/HardwareUtils.cs

```diff
@@ -315,29 +323,40 @@ public void Set333( string value )
             }
         }
 
+        static ConditionalWeakTable<IMachineInfoProvider, Task<MachineID>> generationTable = new ConditionalWeakTable<IMachineInfoProvider, Task<MachineID>>();
 
-        static Task<MachineID>? generateTask;
-
-
-        public static void Init()
+        public static void Init(IMachineInfoProvider machineInfoProvider)
         {
-            generateTask = Task.Factory.StartNew( GenerateMachineID );
+            lock (machineInfoProvider)
```

Comment:
Locking on an object that's publicly accessible isn't ideal...

I wonder though - do we need to support initializing multiple providers? Would a single `Lazy<T>` suffice here?

Regardless I think this is fine as-is.

Reply:
I did consider asking the provider for a lock object [like ICollection/Array/etc. has](https://docs.microsoft.com/en-us/dotnet/api/system.collections.icollection.syncroot?view=net-5.0) but in the end I figured it is probably unnecessary.